Initial Commit For ReSharper Migration Guide (#154)
JetBrains/jetbrains_guide: sites/dotnet-guide/contents/tutorials/resharper-to-rider/rider-debugger-for-resharper-users/index.md

---
type: TutorialStep

> date: 2021-01-29

Reviewer: ```suggestion
date: 2021-03-29
```

title: Debugging For ReSharper Users
technologies: [.net,csharp]
products: [rider,resharper]
topics: [ide]
author: khalidabuhakmeh
subtitle: Learn the basics of the Rider debugger from a Visual Studio + ReSharper user's perspective.
thumbnail: ./thumbnail.png
---

As developers, we try our hardest to avoid introducing bugs and exceptions into our codebase. Unfortunately, regardless of the care we take, mistakes are inevitable. The debugging experience is a stand-out feature of the .NET platform, and it gets even better withing JetBrains Rider. Many of the same features ReSharper users have learned to enjoy are there in the Rider experience, along with some additional UI upgrades that we know our users love.

> ## Main Toolbar and Play Button

Reviewer: ```suggestion
## Main Toolbar and Run Button
```

> One of the first changes developers will notice from Visual Studio is the **play** button's function in the main toolbar. In Visual Studio, the **play** will immediately initiate a debugging session. We may be used to running our applications with the debugger attached, but there are side-effects to getting in the habit of always running in debug mode:

Reviewer: ```suggestion
One of the first changes developers will notice from Visual Studio is the **Run** button in the main toolbar, shown as a _play_ icon. In Visual Studio, the **Run** button will immediately initiate a debugging session. We may be used to running our applications with the debugger attached, but there are side-effects to getting in the habit of always running in debug mode:
```

- Our applications will operate slower due to debugging overhead.
- We are running unoptimized code, never experiencing production code until further down the development lifecycle.
- Utilize more resources during development than is necessary.

> In Rider, users will see a clear functional difference, as the **play** button will only start the current run configuration. To attach the debugger, we can either click the **debug** icon, use the **debug** shortcut, or attach the debugger to an existing process.

Reviewer: ```suggestion
In Rider, users will see a clear functional difference, as the **Run** button will only start the current run configuration. To attach the debugger, we can either click the **Debug** icon, use the **Debug** shortcut, or attach the debugger to an existing process.
```

## Debugging Shortcuts

Many of the same debugging shortcuts are available in Rider, depending on the keymap developers chose during the JetBrains Rider's initial setup.

> - `F5`: Start a new debugging session or resume the current session.
> - `F9`: Set a breakpoint on the current line.
> - `F10`: Stepover current line.
> - `F11`: Step into current line.

Reviewer: ```suggestion
- <kbd>F5</kbd>: Start a new debugging session or resume the current session.
- <kbd>F9</kbd>: Set a breakpoint on the current line.
- <kbd>F10</kbd>: Step over current line.
- <kbd>F11</kbd>: Step into current line.
```

![JetBrains Rider debugger tool window's buttons](./rider-debugger-tool-window-buttons.png)

### Managing Breakpoints

Most developers will be happy managing breakpoints in the editor using the `F9` shortcut or by clicking the space in the gutter. As the complexity of our apps increases, managing breakpoints can help increase our productivity. Below the action buttons in the debugger window, developers will find the **View Breakpoints and Exceptions** button. Here we can see all existing breakpoints, add or modify and conditions, and enable and disable CLR exceptions.

![JetBrains Rider breakpoints tool window](./rider-breakpoints-window.png)

### Locals, Auto, and Watches

Within the Visual Studio UI, **autos**, **locals**, and **watches** are different windows. Within Rider, we can find all variables within the debugger tool window. In addition to seeing the values of variables within the debugger tool window, we can see current values right in the editor with what we call **debugging adornments**.

> [Check out our guide on debugging tips and tricks to learn more about debugging adornments along with other features.](/tutorials/rider-essentials/debugging/)

Reviewer: Will 404
